kliff/uq/bootstrap.py

```diff
@@ -0,0 +1,719 @@
+    """
+
+    def __init__(self, loss: Loss):
+        super().__init__(loss)
+        # Cache the original compute arguments
+        if isinstance(self.calculator, Calculator):
+            self.orig_compute_arguments = [
+                copy.copy(self.calculator.get_compute_arguments())
+            ]
+            self.use_multi_calc = False
+        elif isinstance(self.calculator, _WrapperCalculator):
+            self.orig_compute_arguments = copy.copy(
+                self.calculator.get_compute_arguments(flat=False)
+            )
+            self.use_multi_calc = True
+        self._orig_compute_arguments_identifiers = get_identifiers_from_compute_arguments(
+            self.orig_compute_arguments
+        )
+
+    def generate_bootstrap_compute_arguments(
+        self,
+        nsamples: int,
+        bootstrap_cas_generator_fn: Optional[Callable] = None,
+        **kwargs,
+    ):
+        """
+        Generate bootstrap compute arguments samples. If this function is called multiple,
+        say, K times, then it will in total generate: math: `K \times nsamples` bootstrap
+        compute arguments samples. That is , consecutive call of this function will append
+        the generated compute arguments samples.
+
+        Parameters
+        ----------
+        nsamples: int
+            Number of bootstrap samples to generate.
+        bootstrap_cas_generator_fn: callable ``fn(nsamples, **kwargs)`` (optional)
+            A function to generate bootstrap compute argument samples. The default
+            function combine the compute arguments across all calculators and do sampling
+            with replacement from the combined list. Another possible convention is to
+            do sampling with replacement on the compute arguments list of each calculator
+            separately, in which case a custom function needs to be defined and used.
+        kwargs: dict
+            Additional keyword arguments to ``bootstrap_cas_generator_fn``.
+        """
+        # Function to generate bootstrap configurations
+        if bootstrap_cas_generator_fn is None:
+            bootstrap_cas_generator_fn = bootstrap_cas_generator_empirical
+            kwargs = {"orig_cas": self.orig_compute_arguments}
+        self._generate_bootstrap_compute_arguments(
+            nsamples, bootstrap_cas_generator_fn, **kwargs
+        )
+
+    def save_bootstrap_compute_arguments(self, filename: str):
+        """
+        Export the generated bootstrap compute arguments as a json file. The json file
+        will contain the identifier of the compute arguments for each sample.
+
+        Parameters
+        ----------
+        filename: str or Path
+            Where to export the bootstrap compute arguments samples
+        """
+        self._save_bootstrap_compute_arguments(
+            filename, get_identifiers_from_compute_arguments
+        )
+
+    def load_bootstrap_compute_arguments(self, filename: str):
```

Comment:
Is it important to know what configs are in each bootstrapped dataset? If not, in `save_bootstrap_compute_arguments`, the cas can be dumped as a pickle, instead of json? 

If it is important, we can probably write two files, one is what you are doing now, the other is the pickle. Then when loading them back, we can simply check their identifiers agree with each other. This avoids creating the compute arguments when loading them back. 

A third option, the identifiers and real compute arguments can be easily combined in a data structure, say a dictionary, and save to pickle. But I prefer the second approach above, in case our code base changes, we can still read from a simple json file what are the bootstrapped datasets.